### ISSUE TITLE
Keycloak SAML - There is no AttributeStatement on the Response

### ISSUE THREAD

**Issue by sgohl** [getsentry/self-hosted#3062]
### Self-Hosted Version

24.6.0.dev0

### CPU Architecture

x86_64

### Docker Version

24.0.5

### Docker Compose Version

2.16.0

### Steps to Reproduce

Precondition:

- fresh sentry installation from master branch - only superuser created by `install.sh` - organization name "sentry" - everything is factory defaults
- fresh keycloak installation (`24.0.4`) - with ldap/ad federation

I followed the instructions from here: https://yyhh.org/blog/2020/10/how-to-setup-saml2-authentication-on-sentry-with-keycloak/

With the difference that keycloak metadata url now is without the  `auth` before `realms` in uri.

I tried also with `Name ID format` "username" and "email" – no difference. (because https://github.com/getsentry/self-hosted/issues/2743#issuecomment-1916180615)

I tried ACS/SLS uris both with and without organization name at the end of uri – no difference. (because https://github.com/getsentry/self-hosted/issues/1344#issuecomment-1085827928)

For mappers, I have set them in the default created dedicated scope for the client in tab "Client scopes" (this was reorganized in new keycloak version).  (I also tried creating a new scope and add this mapper, and then add the scope (type default) to the client – no difference)
I deleted the `roles_list` mapping and only added predefined "x500 email"  with `SAML Attribute Name` changed to `user_email`

Other steps:

- Client signature required (tab "Keys") -> OFF
- Sign documents/assertions -> OFF
- Valid redirect/post-logout URIs -> *

Keycloak and Sentry running via docker on same host, but communication via https over central SSL terminated reverse proxy (HAProxy) using real LetsEncrypt cert.

- Client ID: https://sentry.mydomain.dev/saml/metadata/sentry/
- Metadata URL: https://keycloak.mydomain.dev/realms/master/protocol/saml/descriptor
- Metadata XML: https://gist.github.com/sgohl/bff1754e3cdaea599dd3b9ea751c2a2e
- Keycloak Client configuration: https://gist.github.com/sgohl/b0c8ee0a1b0ba84a62a11e6a84912ca0

Both IdP User ID and User Email: `user_email`
Both First Name and Last Name empty for now (elimination testing).

Click "Save Attributes" -> redirect to Keycloak/https works ->  Login as admin -> Redirect back to Sentry/https -> Error:
![image](https://github.com/getsentry/self-hosted/assets/8873017/99f3a6e0-905c-45f1-8bc0-853a3318d455)

I used a SAML tracer browser plugin, and indeed, the AttributeStatement is missing in the saml reposonse: https://gist.github.com/sgohl/9f6504746b43c518be4f54d7889ae6fd

I'm stuck debugging. There are no logs in keycloak (level DEBUG) or sentry-web. It seems, keycloak just omits this statement because it has no reason to include it which is no error per se. 
So I think, something is missing in keycloak configuration, but I can't figure what.

I've found and read these issues:
- https://github.com/getsentry/self-hosted/issues/2743
- https://github.com/getsentry/self-hosted/issues/1571
- https://github.com/getsentry/self-hosted/issues/1344
- https://github.com/getsentry/sentry/issues/71115

and applied any conclusion I found, but for me, nothing worked.

I understand, this is most likely not a sentry issue itself but rather a keycloak configuration, but since the keycloak configuration highly depends on what sentry expects, I feel it belongs to here.

Highly appreciate any help!
Perhaps someone could share their anonymized keycloak client configuration and share how a correct saml response with the AttributeStatement looks like!

### Expected Result

AttributeStatement on the Response not missing - as expected

### Actual Result

Authentication error: SAML SSO fehlgeschlagen, There is no AttributeStatement on the Response

**Comment by azaslavsky** [comment]
Are you seeing any errors in your `sentry-web-1` server logs when you try to do this?

This does seem to be some unique keyclock + sentry combination error, hard to speculate what might be causing it though. We have a specific way of doing it on prod that unfortunately doesn't overlap much with the self-hosted flow.

**Comment by madalinignisca** [comment]
Create in Client Scopes a new entry, name it Sentry for example, or how you wish if you will reuse it.

Create mapper with x500 email, firstname and lastname. Another custom mapper will be for username.

Make sure to make all attributes very simple:
<img width="490" alt="Screenshot 2024-05-21 at 17 27 32" src="https://github.com/getsentry/self-hosted/assets/619561/1f9ba6dd-39d1-48a3-b725-0f54fc7f863a">

Now make sure the Sentry client will use the Client Scope.

This means that Keycloak will return in the response the attributes expected by Sentry, which you must match on the sentry configuration form.

This happens on recent Sentry with lastest 24.x of Keycloak. I spent like 4-5h trying all combination possible from all github issues and referenced blog posts. None work anymore.

**Comment by sgohl** [comment]
Hi, thanks for your response!

So you made it working?

I removed the mappers from the dedicated builtin scope and created a new client scope "sentry" (and added to the client as type `Default`) with mappers for email and `username` (I didn't know that `username` needs a mapper, too?), and tried with `NameFormat` as `Basic` as well as with default urn. Same with firstName and lastName. None worked.

I tried with naming the email/user attribute `email` and `user_email` but both diodn't work (is the name relevant?)

Again tried client/`NameFormat` as "username" and "email".

@madalinignisca 
Could you possibly share the relevant parts of your client configuration (via keycloak export from the Action dropdown) via gist for a day or two so I could compare to mine?

**Comment by azaslavsky** [comment]
It is very surprising that you are getting errors in the UI, but nothing in the web logs. If you open your chrome devtools when you see that error (from the original post), are there any notable 4xx or 5xx codes?

**Comment by sgohl** [comment]
no 4xx or 5xx requests in network inspector, no adblocker, no relevant bad logs in browser console, neither firefox nor chrome

as far as I see, there is a first POST request to sentry which responds with 302 to keycloak, which responds with 200.
Finally a 204 No Content Response from sentry

![image](https://github.com/getsentry/self-hosted/assets/8873017/071cbc93-c1f9-4faf-9ad1-18844cb67f51)

**Comment by madalinignisca** [comment]
@sgohl export.json

```
{
  "clientId": "https://[sentry_host_name]/saml/metadata/[sentry_org]/",
  "name": "",
  "description": "",
  "rootUrl": "",
  "adminUrl": "",
  "baseUrl": "",
  "surrogateAuthRequired": false,
  "enabled": true,
  "alwaysDisplayInConsole": false,
  "clientAuthenticatorType": "client-secret",
  "redirectUris": [
    "https://[sentry_host_name]/*"
  ],
  "webOrigins": [],
  "notBefore": 0,
  "bearerOnly": false,
  "consentRequired": false,
  "standardFlowEnabled": true,
  "implicitFlowEnabled": false,
  "directAccessGrantsEnabled": true,
  "serviceAccountsEnabled": false,
  "publicClient": true,
  "frontchannelLogout": false,
  "protocol": "saml",
  "attributes": {
    "saml.assertion.signature": "false",
    "saml.force.post.binding": "true",
    "saml_single_logout_service_url_post": "https://[sentry_host_name]/saml/sls/",
    "saml.encrypt": "false",
    "saml_assertion_consumer_url_post": "https://[sentry_host_name]/saml/acs/",
    "saml.server.signature": "true",
    "saml.server.signature.keyinfo.ext": "false",
    "saml.signing.certificate": "[certificate]",
    "saml.artifact.binding.identifier": "[some-uuid]",
    "saml.artifact.binding": "false",
    "saml.signature.algorithm": "RSA_SHA256",
    "saml_force_name_id_format": "false",
    "saml.client.signature": "false",
    "saml.authnstatement": "true",
    "display.on.consent.screen": "false",
    "saml_name_id_format": "username",
    "saml.signing.private.key": "[key]",
    "saml.allow.ecp.flow": "false",
    "saml_signature_canonicalization_method": "http://www.w3.org/2001/10/xml-exc-c14n#",
    "saml.onetimeuse.condition": "false",
    "saml.server.signature.keyinfo.xmlSigKeyInfoKeyNameTransformer": "NONE"
  },
  "authenticationFlowBindingOverrides": {},
  "fullScopeAllowed": true,
  "nodeReRegistrationTimeout": -1,
  "defaultClientScopes": [
    "Sentry"
  ],
  "optionalClientScopes": [],
  "access": {
    "view": true,
    "configure": true,
    "manage": true
  }
}
```

**Comment by madalinignisca** [comment]
@sgohl make sure you have the mappers right and used in the client!

I have this, and in each one, there is the `SAML Attribute Name` expected by Sentry and matching your configuration.

<img width="733" alt="Screenshot 2024-05-26 at 10 40 15" src="https://github.com/getsentry/self-hosted/assets/619561/6cfda64a-4d55-498b-8910-c8e523f9ecac">

**Comment by madalinignisca** [comment]
The problem is not with Sentry. Sentry works fine with custom SAML2 providers.

The problem is lack of information anywhere on the internet how to setup Keycloak for any usecase. There are many tutorials for some popular integrations, but nothing relevant for majority. Probably most companies end with some paid saas solution (AD included), as those are preconfigured for most case scenarios.

Keycloak is a bare bones solution for which who manages is fully responsible to know expectations of CLIENT and setup all required scenarios.

If you are trying Keycloak and you don't have anything else setup in it, maybe a more simple SAML provider alternative would better help, like `lemonldap-ng` or `simplesamlphp`, which compared to Keycloak are human level stuff, where Keycloak is enterprise level made to justify time consumed to configure it.

**Comment by sgohl** [comment]
I got it.  It had nothing to do with the mappers, they were fine.

My misconception was: I switched between the two error messages:

1) There is no AttributeStatement on the Response
2) No Signature found. SAML Response rejected

while I thought 1) would be a step further. But it wasn't. Inspecting the SAML response showed the AttributeStatements for all mappers.  In contrary to this comment https://github.com/getsentry/self-hosted/issues/1571#issuecomment-1187012447  that said, fixing the mappers fixed the issue, for me it was the following problem:


*EVERY*  tutorial out there said: Disable the Signing stuff (Sign documents, Sign assertions)

But actually that is what was missing. Now i have:

![image](https://github.com/getsentry/self-hosted/assets/8873017/1f7f1980-5b0c-496d-a8eb-80a3cc28d4d8)

and it works.
